certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,160 @@
 using SafeHarness as safeContract;
+using GuardianStorageHarness as guardianStorageContract;
+
+definition SENTINEL() returns address = 1;
 
 methods {
+    // Social Recovery Module Functions
+    function isGuardian(address, address) external returns (bool) envfree;
+    function guardiansCount(address) external returns (uint256) envfree;
+    function threshold(address) external returns (uint256) envfree;
+
+    // Guardian Storage Functions
+    function guardianStorageContract.countGuardians(address) external returns (uint256) envfree;
+
     // Safe Functions
-    function safeContract.isModuleEnabled(address) external returns (bool) envfree;
+    function safeContract.isModuleEnabled(address module) external returns (bool) envfree;
+    function safeContract.isOwner(address owner) external returns (bool) envfree;
+
+    // Wildcard Functions (Because of use of ISafe interface in Social Recovery Module)
+    function _.isModuleEnabled(address module) external => safeIsModuleEnabled(calledContract, module) expect bool ALL; // `calledContract` is a special variable.
+    function _.isOwner(address owner) external => sumarizeSafeIsOwner(calledContract, owner) expect bool ALL;
+}
+
+// A summary function to check if a module is enabled in the Safe contract.
+function safeIsModuleEnabled(address callee, address module) returns bool {
+    assert callee == safeContract;
+    return safeContract.isModuleEnabled(module);
 }
 
-// A setup function that requires Safe contract to enabled the Social Recovery
-// Module.
+// A summary function to check if an address is an owner in the Safe contract.
+function sumarizeSafeIsOwner(address callee, address owner) returns bool {
+    assert callee == safeContract;
+    return safeContract.isOwner(owner);
+}
+
+// A setup function that requires Safe contract to enabled the Social Recovery Module.
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
```

Comment:
Suggestion: We could add a `requireGuardiansLinkedListIntegrity` helper function instead of repeating `require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract)` in every rule.

This way, once we have the required invariants setup, we can update the function in one place?

Also - I _think_ we will need this even when we have the `guardianstorage` invariants, since invariants can only apply to the "main" contract from a CVL spec, which is `SafeSocialRecovery` here.

Reply:
Function added 👍🏾 

In the call, Jochen mentioned importing invariants which are proved already in other file. We can definitely give that a try. For now, I have created this issue: #24